pkg/internal/controller/controller.go

```diff
@@ -144,6 +220,39 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 	return *c.LeaderElected
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+	if c.EnableWarmup == nil || !*c.EnableWarmup {
+		return nil
+	}
+
+	// Hold the lock to avoid concurrent access to c.startWatches with Start() when calling
+	// startEventSources
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	err := c.startEventSources(ctx)
+	c.warmupResultChan <- err
+
+	return err
+}
+
+// WaitForWarmupComplete returns true if warmup has completed without error, and false if there was
+// an error during warmup. If context is cancelled, it returns true.
+func (c *Controller[request]) WaitForWarmupComplete(ctx context.Context) bool {
+	if c.EnableWarmup == nil || !*c.EnableWarmup {
+		return true
+	}
+
+	warmupError, ok := <-c.warmupResultChan
+	if !ok {
+		// channel closed unexpectedly
+		return false
+	}
+
+	return warmupError == nil
```

Comment:
So what happens if we hit the error case?

Seems like we then never signal ready on startReadyCh, which means the runnableGroup.Start func only returns if the ctx is cancelled? (which if I see correctly is equivalent to it only returns if the ctx passed into controllerManager.Start returns?)